workspaces/arborist/lib/node.js

```diff
@@ -1344,9 +1344,99 @@ class Node {
+    for (const edge of this.edgesIn) {
+      if (newOverrideSet) {
+        newOverrideSet = this.findSpecificOverrideSet(edge.overrides, newOverrideSet)
+      } else {
+        newOverrideSet = edge.overrides
+      }
+    }
+    if (this.overrides == newOverrideSet) {
+      return false
+    }
+    this.overrides = newOverrideSet
+    if (this.overrides) {
+      // Optimization: if there's any override set at all, then no non-extraneous node has an empty override set. So if we temporarily have no
+      // override set (for example, we removed all the edges in), there's no use updating all the edges out right now. Let's just wait until
+      // we have an actual override set later.
+      this.recalculateOutEdgesOverrides()
+    }
+    return true
+  }
+
+  // This logic isn't perfect either. When we have two edges in that have different override sets, then we have to decide which set is correct.
+  // This function assumes the more specific override set is applicable, so if we have dependencies A->B->C and A->C
+  // and an override set that specifies what happens for C under A->B, this will work even if the new A->C edge comes along and tries to change
+  // the override set.
+  // The strictly correct logic is not to allow two edges with different overrides to point to the same node, because even if this node can satisfy
+  // both, one of its dependencies might need to be different depending on the edge leading to it.
+  // However, this might cause a lot of duplication, because the conflict in the dependencies might never actually happen.
+  updateNodeOverrideSet (otherOverrideSet) {
+    if (this.overrides == otherOverrideSet) {
+      return false
+    }
+    if (!this.overrides) {
+      this.overrides = otherOverrideSet
+      this.recalculateOutEdgesOverrides()
+      return true
+    }
+    let newOverrideSet = this.findSpecificOverrideSet(this.overrides, otherOverrideSet)
+    if (!newOverrideSet) {
+      // This is an error condition. We can only get here if the new override set is in conflict with the existing.
+      return false
+    } else {
+      if (this.overrides != newOverrideSet) {
+        this.overrides = newOverrideSet
+        this.recalculateOutEdgesOverrides()
+        return true
+      }
+      return false
+    }
+  }
+  
+  deleteEdgeIn (edge) {
+    this.edgesIn.delete(edge)
     if (edge.overrides) {
-      this.overrides = edge.overrides
+      this.updateNodeOverrideSetDueToEdgeRemoval(edge.overrides)
+    }
+  }
+
+  addEdgeIn (edge) {
+    // We need to handle the case where the new edge in has an overrides field which is different from the current value.
+    // Assuming there are any overrides at all, the overrides field is never undefined for any node at the end state of the tree.
+    // So if the new edge's overrides is undefined it will be updated later. So we can wait with updating the node's overrides field.
+    if (edge.overrides && this.overrides != edge.overrides) {
```

Comment:
We tend to avoid `!=` and `==`.  As far as I can tell we only have one in the constructor of edges, and even that should probably be refactored to be removed.

This comparator also seems overly optimistic, as an edge's overrides reduce down to the single rule which applies for that edge.

Reply:
You're right. I'll add a proper isEquals function.